dask/array/tests/test_array_function.py

```diff
@@ -104,3 +104,103 @@ def test_array_function_cupy_svd():
     assert_eq(u, u_base)
     assert_eq(s, s_base)
     assert_eq(v, v_base)
+
+
+@pytest.mark.skipif(missing_arrfunc_cond, reason=missing_arrfunc_reason)
+@pytest.mark.parametrize('func', [
+    lambda x: np.concatenate([x, x, x]),
+    lambda x: np.cov(x, x),
+    lambda x: np.dot(x, x),
+    lambda x: np.dstack(x),
+    lambda x: np.flip(x, axis=0),
+    lambda x: np.hstack(x),
+    lambda x: np.matmul(x, x),
+    lambda x: np.mean(x),
+    lambda x: np.stack([x, x]),
+    lambda x: np.sum(x),
+    lambda x: np.var(x),
+    lambda x: np.vstack(x),
+    lambda x: np.linalg.norm(x)])
+def test_unregistered_func(func):
+    def wrap(func_name):
+        """
+        Wrap a function that returns an array.
+        """
+        def wrapped(self, *a, **kw):
+            return type(self)(getattr(self.arr, func_name)(*a, **kw))
+
+        return wrapped
+
+    def dispatch(func_name):
+        """
+        Wrap a function that doesn't return an array.
+        """
+        def wrapped(self, *a, **kw):
+            return getattr(self.arr, func_name)(*a, **kw)
+
+        return wrapped
+
+    def dispatch_property(prop_name):
+        """
+        Wrap a simple property.
+        """
+        @property
+        def wrapped(self, *a, **kw):
+            return getattr(self.arr, prop_name)
+
+        return wrapped
+
+    class EncapsulateNDArray(np.lib.mixins.NDArrayOperatorsMixin):
+        """
+        A class that "mocks" ndarray by encapsulating an ndarray and using
+        protocols to "look like" an ndarray. Basically tests whether Dask
+        works fine with something that is essentially an array but uses
+        protocols instead of being an actual array.
+        """
+        __array_priority__ = 20
+
+        def __init__(self, arr):
+            self.arr = arr[...]
+
+        def __array__(self, **kwargs):
+            return np.asarray(self.arr, **kwargs)
+
+        def __array_function__(self, f, t, arrs, kw):
+            arrs = tuple(arr if not isinstance(arr, type(self)) else arr.arr for arr in arrs)
+            return type(self)(self.arr.__array_function__(f, (np.ndarray,), arrs, kw))
```

Comment:
@shoyer @mhvk This "hack" might be interesting to you. ;) It'll need to be changed, but I couldn't find a better way to do it.

Reply:
I realize it is for testing, but setting `types` to `(np.ndarray,)` will break functions in which another argument also overrides `__array_function__`. Instead, remove `type(self)` from `types`.
Also, if there are other types present, `ndarray.__array_function__(...)` should return `NotImplemented`, so you'd need to pass that on (I also check for `None` though I'm not sure that is relevant for functions - it is for `ufunc.at`).
Not sure if it helps, but the Quantity implementation is at https://github.com/astropy/astropy/blob/master/astropy/units/quantity.py#L1460 - note that I don't have to remove `Quantity` from the types since `Quantity` is an `ndarray` subclass.